lib/data_stores/mysql_advanced_data_store.py

```diff
@@ -0,0 +1,636 @@
+      aff4["args"] = [value, timestamp, subject, predicate]
+      transaction.append(aff4)
+    return transaction
+
+  def _BuildInserts(self, values):
+    subjects = {}
+    attributes = {}
+    aff4 = {}
+
+    subjects["query"] = "INSERT IGNORE INTO subjects (hash, subject) VALUES"
+    attributes["query"] = "INSERT IGNORE INTO attributes (hash, attribute) VALUES"
+    aff4["query"] = "INSERT INTO aff4 (subject_hash, attribute_hash, prefix, \
+timestamp, value) VALUES"
+
+    subjects["args"] = []
+    attributes["args"] = []
+    aff4["args"] = []
+
+    value_sets = zip(values[0::4], values[1::4], values[2::4], values[3::4])
+
+    seen = {}
+    seen["subjects"] = []
+    seen["attributes"] = []
+
+    for (subject, predicate, value, timestamp) in value_sets:
+      if subject not in seen["subjects"]:
+        subjects["args"].extend([subject, subject])
+        seen["subjects"].append(subject)
+      if predicate not in seen["attributes"]:
+        attributes["args"].extend([predicate, predicate])
+        seen["attributes"].append(predicate)
+      prefix = predicate.split(":", 1)[0]
+      aff4["args"].extend([subject, predicate, prefix, timestamp, value])
+
+    subjects["query"] += ", ".join(["(unhex(md5(%s)), %s)"] * (len(subjects["args"]) / 2))
+    attributes["query"] += ", ".join(["(unhex(md5(%s)), %s)"] * (len(attributes["args"]) / 2))
+    aff4["query"] += ", ".join(
+        ["(unhex(md5(%s)), unhex(md5(%s)), %s, %s, %s)"] * (
+            len(aff4["args"]) / 5))
+
+    return [subjects, attributes, aff4]
+
+  def _ExecuteInsert(self, transaction):
+    """Get connection from pool and execute query"""
+    for query in transaction:
+      with self.pool.GetConnection() as cursor:
+        cursor.Execute(query["query"], query["args"])
+
+  def _ResolveRegex(self, subject, predicate_regex, token=None, timestamp=None, limit=None):
+    if limit and limit == 0:
+      return []
+
+    if isinstance(predicate_regex, str):
+      predicate_regex = [predicate_regex]
+
+    results = []
+    seen = set()
+
+    for regex in predicate_regex:
+      query, args = self._BuildQuery("SELECT", subject, regex, timestamp, limit, is_regex=True)
+      rows = self._ExecuteQuery(query, args)
+
+      for row in rows:
+        attribute = row["attribute"]
+        value = self._DecodeValue(attribute, row["value"])
+        results.append((attribute, value, row["timestamp"]))
+        if limit:
+          limit -= 1
+          if limit == 0:
+            return results
+
+    return results
+
+  def _CalculateAttributeStorageTypes(self):
+    """Build a mapping between column names and types."""
+    self.attribute_types = {}
+
+    for attribute in aff4.Attribute.PREDICATES.values():
+      self.attribute_types[attribute.predicate] = (
+          attribute.attribute_type.data_store_type)
+
+  def _EncodeValue(self, value):
+    """Encode the value for the attribute."""
+    if hasattr(value, "SerializeToString"):
```

Comment:
Hm interesting, in the MySQL DS we do try/except. Can you maybe change both here and SQLite so we have some consistency?